keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt

```diff
@@ -15,58 +15,81 @@
  */
 package com.netflix.spinnaker.keel.clouddriver
 
-import com.github.benmanes.caffeine.cache.Cache
+import com.github.benmanes.caffeine.cache.AsyncCache
 import com.github.benmanes.caffeine.cache.Caffeine
 import com.netflix.spinnaker.keel.clouddriver.model.Credential
 import com.netflix.spinnaker.keel.clouddriver.model.Network
 import com.netflix.spinnaker.keel.clouddriver.model.SecurityGroupSummary
 import com.netflix.spinnaker.keel.clouddriver.model.Subnet
 import com.netflix.spinnaker.keel.core.api.DEFAULT_SERVICE_ACCOUNT
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.Executor
 import java.util.concurrent.TimeUnit.HOURS
 import java.util.concurrent.TimeUnit.MINUTES
+import java.util.function.BiFunction
 import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.future.future
 import kotlinx.coroutines.runBlocking
 import retrofit2.HttpException
 
+/**
+ * An in-memory cache for calls against cloud driver
+ *
+ * Caching is implemented using asynchronous caches ([AsyncCache]) because
+ * it isn't safe for a kotlin coroutine to yield inside of the second argument
+ * of a synchronous cache's get method.
+ *
+ * For more details on why using async, see: https://github.com/spinnaker/keel/pull/1154
+ */
 class MemoryCloudDriverCache(
   private val cloudDriver: CloudDriverService
 ) : CloudDriverCache {
 
   private val securityGroupSummariesByIdOrName = Caffeine.newBuilder()
     .maximumSize(1000)
     .expireAfterWrite(1, MINUTES)
-    .build<String, SecurityGroupSummary>()
+    .buildAsync<String, SecurityGroupSummary>()
 
   private val networks = Caffeine.newBuilder()
     .maximumSize(1000)
     .expireAfterWrite(1, HOURS)
-    .build<String, Network>()
+    .buildAsync<String, Network>()
 
   private val availabilityZones = Caffeine.newBuilder()
     .maximumSize(1000)
     .expireAfterWrite(1, HOURS)
-    .build<String, Set<String>>()
+    .buildAsync<String, Set<String>>()
 
   private val credentials = Caffeine.newBuilder()
     .maximumSize(100)
     .expireAfterWrite(1, HOURS)
-    .build<String, Credential>()
+    .buildAsync<String, Credential>()
 
   private val subnetsById = Caffeine.newBuilder()
     .maximumSize(1000)
     .expireAfterWrite(1, HOURS)
-    .build<String, Subnet>()
+    .buildAsync<String, Subnet>()
 
   private val subnetsByPurpose = Caffeine.newBuilder()
     .maximumSize(1000)
     .expireAfterWrite(1, HOURS)
-    .build<String, Subnet>()
+    .buildAsync<String, Subnet>()
 
   override fun credentialBy(name: String): Credential =
     credentials.getOrNotFound(name, "Credentials with name $name not found") {
       cloudDriver.getCredential(name, DEFAULT_SERVICE_ACCOUNT)
     }
 
+  /**
+   * Convert a suspending function to a [BiFunction] that [AsyncCache.get] expects as its second argument
+   *
+   * Uses [Dispatchers.IO] because it assumes that [block] makes I/O calls
```

Comment:
To the best of my understanding this is right. The cache will be primed by HTTP calls to CloudDriver.